docs/docs/higher-order-components.md

```diff
@@ -0,0 +1,491 @@
+```
+
+Usually, HOCs accept additional arguments. In this example from Relay, a config object is used to specify a component's data dependencies:
+
+```js
+const CommentWithRelay = Relay.createContainer(Comment, config);
+```
+
+By far, the most common signature for HOCs looks like this:
+
+```js
+// React Redux's `connect`
+const ConnectedComment = connect(commentSelector, commentActions)(Comment);
+```
+
+*What?!* If you break it apart, it's easier to see what's going on.
+
+```js
+// connect is a function that returns another function
+const enhance = connect(commentListSelector, commentListActions);
+// The returned function is an HOC, which returns a component that is connected
+// to the Redux store
+const ConnectedComment = enhance(CommentList);
+```
+In other words, `connect` is a higher-order function that returns a higher-order component!
+
+This form may seem confusing or unnecessary, but it has a useful property. Single-argument HOCs like the one returned by the `connect` function have the signature `Component => Component`. Functions whose output type is the same as its input type are really easy to compose together.
+
+```js
+// Instead of doing this...
+const EnhancedComponent = connect(commentSelector)(withRouter(WrappedComponent))
+
+// ... you can use a function composition utility
+// compose(f, g, h) is the same as (...args) => f(g(h(...args)))
+const enhance = compose(
+  // These are both single-argument HOCs
+  connect(commentSelector),
+  withRouter
+)
+const EnhancedComponent = enhance(WrappedComponent)
+```
+
+(This same property also allows `connect` and other enhancer-style HOCs to be used as decorators, an experimental JavaScript proposal.)
+
+The `compose` utility function is provided by many third-party libraries including lodash (as [`lodash.flowRight`](https://lodash.com/docs/#flowRight)), [Redux](http://redux.js.org/docs/api/compose.html), and [Rambda](http://ramdajs.com/docs/#compose).
+
+### Optional: Wrap the Display Name for Easy Debugging
+
+The container components created by HOCs show up in the React DevTools like any other component. To ease debugging, choose a display name that communicates that it's the result of an HOC.
+
+The most common technique is to wrap the display name of the wrapped component. So if your higher-order component is named `withSubscription`, and the wrapped component's display name is `CommentList`, use the display name `WithSubscription(CommentList)`:
+
+```js
+function withSubscription(WrappedComponent) {
+  class WithSubscription extends React.Component {/* ... */}
+  WithSubscription.displayName = `WithSubscription(${getDisplayName(WrappedComponent)})`;
+  return WithSubscription;
+}
+
+function getDisplayName(WrappedComponent) {
+  return WrappedComponent.displayName || WrappedComponent.name || 'Component';
+}
+```
+
+
+## Caveats
+
+We consider HOCs an advanced technique because they come with a few caveats that aren't immediately obvious if you're new to React.
```

Comment:
Just to shorten up the prose how about

"We consider HOCs an advanced technique because they" -> "Higher-order components"